src/ppm.rs

```diff
@@ -0,0 +1,499 @@
+    Validity(&'static str),
+
+    /// PCP error.
+    #[error("ppm error: pcp error: {0}")]
+    Pcp(#[from] PcpError),
+
+    /// Type error.
+    #[error("ppm error: type error: {0}")]
+    Type(#[from] TypeError),
+
+    /// PRNG error.
+    #[error("ppm error: prng error: {0}")]
+    Prng(#[from] PrngError),
+
+    /// Calling get_random() returned an error.
+    #[error("ppm error: getrandom error: {0}")]
+    GetRandom(#[from] getrandom::Error),
+}
+
+/// A share of an input or proof for Prio.
+#[derive(Clone, Debug)]
+pub enum Share<F: FieldElement> {
+    /// An uncompressed share, typically sent to the leader.
+    Leader(Vec<F>),
+
+    /// A compressed share, typically sent to the helper. The first parameter is a seed for a
+    /// pseudorandom generator
+    Helper(Vec<u8>, usize),
+}
+
+impl<F: FieldElement> TryFrom<Share<F>> for Vec<F> {
+    type Error = PpmError;
+
+    fn try_from(share: Share<F>) -> Result<Self, PpmError> {
+        match share {
+            Share::Leader(data) => Ok(data),
+            Share::Helper(seed, length) => {
+                let mut prng = Prng::try_from_seed(&seed)?;
+                Ok((0..length).map(|_| prng.next().unwrap()).collect())
+            }
+        }
+    }
+}
+
+/// The message sent by the client to each aggregator. This includes the client's input share and
+/// the initial message of the input-validation protocol.
+#[derive(Clone, Debug)]
+pub struct Upload<F: FieldElement> {
+    /// The input share.
+    pub input_share: Share<F>,
+
+    /// The proof share.
+    pub proof_share: Share<F>,
+
+    /// The sum of the joint randomness seed shares sent to the other aggregators.
+    pub joint_rand_seed_hint: Vec<u8>,
+
+    /// The blinding factor, used to derive the aggregator's joint randomness seed share.
+    pub blind: Vec<u8>,
+}
+
+fn assert_safe() {
+    // TODO(cjpatton) Replace this hack with a more consistent API for the Prio ciphersuite.Prng
+    // should be constructed with an implementation of the following trait:
+    //
+    // pub trait Cipher {
+    //    // Panics if `seed.len() != StreamCipher::KEY_LEN`.
+    //    fn new_stream_cipher(seed: &[u8]) -> StreamCipher;
+    //
+    //    // Panics if `Hasher::OUT_LEN != StreamCipher::KEY_LEN`.
+    //    fn new_hasher() -> Hasher;
+    // }
+    //
+    // pub trait StreamCipher {
+    //    const KEY_LEN: usize;
+    //
+    //    /// Fill `out` with the next `out.len()` bytes of the stream cipher.
+    //    fn fill(out: &mut [u8]);
+    // }
+    //
+    // pub trait Hasher<const OUT_LEN: usize> {
+    //    const OUT_LEN: usize;
+    //
+    //    /// Appends `input` to the hash input.
+    //    fn update(input: &[u8]);
+    //
+    //    /// Panics if `out.len() != Self::OUT_LEN`.
+    //    fn finalize(out: &mut [u8]);
+    // }
+    const_assert_eq!(blake3::KEY_LEN, SEED_LENGTH);
+    const_assert_eq!(blake3::OUT_LEN, SEED_LENGTH);
+}
+
+#[derive(Clone)]
+struct HelperShare {
```

Comment:
nit: A module-private struct is perfectly adequate, but if you want to hide it even harder you could put the definition inside `fn upload`.

Reply:
Nope, I'm fine with this being defined at the module level.